### PR TITLE
Fix max nan in GPU and Clang

### LINKED ISSUE #862: `.max(Tensor([ 1, nan]))` should be nan
`Tensor([1, np.nan]).numpy().max()` -> `np.nan`
`Tensor([1, np.nan]).max().numpy()` -> `array([1.], dtype=float32)`

Wrong output shape aside, it should be `nan`, not `1`.

### PR DESCRIPTION
**Pull request by uncommonSensor**
* added prefix with max func to GPU CLProgram
* added nan check to GPU and Clang prefix
* unskipped CI max nan check for GPU and Clang
* Fixes #862

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/runtime/ops_clang.py              1               1
tinygrad/runtime/ops_gpu.py                1               0
------------------------------------------------------------
total                                      2               1
------------------------------------------------------------
lines added in the tinygrad folder: 1
```

**Comment by geohot** [comment]
Is this how torch does it? What's the performance implication?

**Comment by uncommonSensor** [comment]
Yes, this is how torch handles it. I created a similar test with torch and it passed (returns nan):

```
GPU=1 python3 test/test_ops.py TestOps.test_max_torch_inf (__main__.TestOps) ... ok

def test_max_inf(self):
    n = torch.max(torch.FloatTensor([1, float("nan")])).numpy()
    assert math.isnan(n.item()), f"{n.item()} is not nan"
```
It's worth noting that pytorch also has **fmax **which will return the higher value even with nan.

checking performance now

File: tinygrad/runtime/ops_gpu.py [hunk 1]
```diff
@@ -50,6 +50,7 @@ def _copyout(self, x: np.ndarray):
 
 class CLProgram:
   def __init__(self, name:str, prg:str, binary=False, argdtypes=None, options=None):
+    if not binary: prg = '#define max(x,y) ((isnan(x) || isnan(y)) ? NAN : ((x > y) ? x : y))\n' + prg
```
[the remark this answers is not in the document]
Comment (uncommonSensor) [reply tinygrad/runtime/ops_gpu.py:53]: no it does not actually

**Comment by uncommonSensor** [comment]
Pytorch max() will return a nan if any element in tensor is nan. There is also fmax() which is faster and will return the higher element if there is a nan. max() is utilized to avoid potential issues caused by invalid calculations. fmax is used when performance is more important.

**Pytorch example:**
```
import torch
a = (torch.tensor([1.0, float('nan')], device='cuda')).max()                                                       
print(a)
tensor(nan, device='cuda:0')**
```

Here are some benchmarks with 3090 GPU:

**THIS PATCH:**
```
Hello.
ran model in 496.13 ms
sync in 0.11 ms
 I
ran model in 647.27 ms
sync in 0.12 ms
'
ran model in 358.04 ms
sync in 0.12 ms
m
ran model in 357.66 ms
sync in 0.11 ms
 a
ran model in 357.37 ms
sync in 0.12 ms
 
ran model in 366.11 ms
sync in 0.14 ms
2
ran model in 359.40 ms
sync in 0.12 ms
0
ran model in 358.46 ms
sync in 0.11 ms
 year
ran model in 358.98 ms
sync in 0.11 ms
 old
ran model in 368.41 ms
sync in 0.13 ms
```

**MASTER:**
```
Hello.
ran model in 491.97 ms
sync in 0.12 ms
 I
ran model in 644.98 ms
sync in 0.12 ms
'
ran model in 359.18 ms
sync in 0.12 ms
m
ran model in 360.02 ms
sync in 0.12 ms
 a
ran model in 360.01 ms
sync in 0.15 ms
 
ran model in 368.71 ms
sync in 0.14 ms
2
ran model in 361.87 ms
sync in 0.12 ms
0
ran model in 360.93 ms
sync in 0.13 ms
 year
ran model in 361.81 ms
sync in 0.12 ms
 old
ran model in 371.50 ms
sync in 0.13 ms
```

flammit  made a good point on the original ticket regarding the different max() behaviors of each backend as well.

File: tinygrad/runtime/ops_gpu.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (uncommonSensor) [reply tinygrad/runtime/ops_gpu.py:53]: I created a  test to check how nan is handled (on master, as well as this patch). If my tests are correct then it seems GPU max already handles nan for float4 this way:

```
def test_max_float4_nan(self):
  n = Tensor(np.array([[1.0, 2.0, np.nan, 4.0], [5.0, 6.0, 7.0, 8.0]], dtype=np.float32)).max().numpy()
  print("Float 4 nan: ", math.isnan(n.item()))
  assert math.isnan(n.item()), f"tinygrad max: {n.item()} is not nan"
```

`Master:` Float 4 nan:  True
`This Patch:` Float 4 nan:  True

Although max does not seem to handle nan with float 4 with or without my change to 'ops_clang.py'

**Comment by geohot** [comment]
Running llama will not test the performance of max. You have to test it explicitly

**Comment by uncommonSensor** [comment]
Here's just the max test:

GPU=1 python3 test/test_speed_v_torch.py TestSpeed.test_max

**THIS PATCH:**
 ```
max   096x 4096    3.71 ms (    4.53 GFLOPS    18.11 GB/s) in torch,    9.52 ms (    1.76 GFLOPS     7.05 GB/s) in tinygrad,    2.57x slower      16.78 MOPS    67.11 MB
.
----------------------------------------------------------------------
Ran 1 test in 0.728s
OK
```
**MASTER:**
```
 max   4096x 4096    3.70 ms (    4.54 GFLOPS    18.16 GB/s) in torch,    9.53 ms (    1.76 GFLOPS     7.05 GB/s) in tinygrad,    2.58x slower      16.78 MOPS    67.11 MB
.
----------------------------------------------------------------------
Ran 1 test in 0.699s
OK
```

**Comment by geohot** [comment]
Can this be done in tensor.py instead? Does it have to change the backends?

**Comment by uncommonSensor** [comment]
I think so, I've done some brief investigating on that. Looking into it.

**Comment by uncommonSensor** [comment]
After further investigation, making changes to the Tensor class does not seem like the best route. Checking nan without realizing the tensors (and using the CPU), doesn't seem feasible. In order to match Tinygrad max() to Pytorch max (returning nan if the tensor's contain nan), the best approach seems to be to make changes to the backends of the accelerators."

**Comment by uncommonSensor** [comment]
I've got a branch that is checking for nan in max without realizing the tensor. Working out some kinks at the moment.

**Pull request closed without merging**